File: backend/src/application/usecase/LoginUserUseCase.go
package usecase

import (
	"github.com/marcosfrancomarinho/go-first-project/src/application/dto"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/entities"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/interfaces"
	"github.com/marcosfrancomarinho/go-first-project/src/domain/valuesobject"
)

type LoginUserUseCase struct {
	encryptor         interfaces.PasswordEncryptor
	userAuthenticator interfaces.UserAuthenticator
	findorUser        interfaces.FindorUser
}

func NewLoginUserUseCase(
	encryptor interfaces.PasswordEncryptor,
	userAuthenticator interfaces.UserAuthenticator,
	findorUser interfaces.FindorUser,
) *LoginUserUseCase {
	return &LoginUserUseCase{encryptor: encryptor, userAuthenticator: userAuthenticator, findorUser: findorUser}
}

func (l *LoginUserUseCase) Login(payload *dto.RequestLoginUserDTO) (*dto.ResponseLoginUserDTO, error) {
	email, err := valuesobject.NewEmail(payload.Email)
	if err != nil {
		return nil, err
	}

	password, err := valuesobject.NewPassword(payload.Password)
	if err != nil {
		return nil, err
	}

	userLogin := entities.NewUserLogin(email, password)

	foundUser, err := l.findorUser.FindByEmail(userLogin)
	if err != nil {
		return nil, err
	}

	if err := l.encryptor.ValidatePassword(password.GetValue(), foundUser.GetPassword()); err != nil {
		return nil, err
	}

	token, err := l.userAuthenticator.GenerateToken(foundUser)
	if err != nil {
		return nil, err
	}

	return &dto.ResponseLoginUserDTO{
		Name:    foundUser.GetName(),
		Message: "usuario logado com sucesso",
		Token:   token.GetValue(),
	}, nil
}

File: frontend/src/application/usecase/LoginUserUseCase.ts
import { User } from '../../domain/entities/User';
import type { HttpPostClient } from '../../domain/gateway/HttpPostClient';
import type { StorageClient } from '../../domain/gateway/StorageClient';
import { Email } from '../../domain/valuesobject/Email';
import { Password } from '../../domain/valuesobject/Password';
import type { RequestLoginUserDTO } from '../dto/RequestLoginUserDTO';
import type { ResponseLoginUserDTO } from '../dto/ResponseLoginUserDTO';

export class LoginUserUseCase {
  public constructor(
    private httpClient: HttpPostClient,
    private storageClient: StorageClient,
    private path: string,
    private key: string
  ) {}

  public async login(payload: RequestLoginUserDTO): Promise<ResponseLoginUserDTO> {
    const email: Email = Email.create(payload.email);
    const password: Password = Password.create(payload.password);
    const user: User = User.create(email, password);
    const authenticationResponse = await this.httpClient.post<ResponseLoginUserDTO>(
      this.path,
      user.getCredentialsForAuthentication()
    );
    this.storageClient.set(this.key, { name: authenticationResponse.name, token: authenticationResponse.token });
    return authenticationResponse;
  }
}
